celerpy/visualize.py

```diff
@@ -0,0 +1,352 @@
+                bin_file=Path(f.name),
+                image=image,
+                **kwargs,
+            )
+            result = process.communicate_model(
+                self.process, inp, model.TraceOutput
+            )
+            img = f.read()
+
+        # Cache the geometry names and ensure trace has them
+        if geometry is None:
+            geometry = result.trace.geometry
+            assert geometry is not None
+            volumes = self.volumes.setdefault(geometry, [])
+        if not volumes:
+            assert isinstance(volumes, list)
+            assert result.volumes
+            # XXX : erase pointer names (for now?)
+            volumes[:] = [_re_ptr.sub("", v) for v in result.volumes]
+        else:
+            result.volumes = volumes
+
+        # Reinterpret the image as a correctly shaped Numpy array
+        npimg = np.frombuffer(img, dtype=np.int32).reshape(result.image.dims)
+        self.image = result.image
+
+        return (result, npimg)
+
+    def close(self, *, timeout=0.25):
+        """Cleanly exit the ray trace loop, returning run statistics if possible."""
+        result = process.communicate(self.process, json.dumps(None))
+        try:
+            self.process.wait(timeout=timeout)
+        except TimeoutExpired:
+            pass
+        result = result or process.close(self.process, timeout=timeout)
+        try:
+            result = json.loads(result)
+        except json.JSONDecodeError:
+            pass
+        return result
+
+
+class ReverseIndexDict(collections.defaultdict):
+    """Manage a two-way mapping of integers and another type."""
+
+    def __init__(self, volume_list):
+        self.vlist = volume_list
+
+    def __missing__(self, key):
+        self[key] = result = len(self)
+        self.vlist.append(key)
+        assert len(self.vlist) == len(self)
+        return result
+
+
+LabeledAxis = collections.namedtuple("LabeledAxis", ["label", "lo", "hi"])
+LabeledAxes = collections.namedtuple("LabeledAxes", ["x", "y"])
+
+
+class IdNorm(BoundaryNorm):
+    """Map IDs to consecutive integers for rendering.
+
+    This provides literally a factor of 1000 speedup over using BoundaryNorm
+    (when trying to plot an image with 3k compositions).
+    """
+
+    # Note: we inherit from BoundaryNorm because there's some special-casing
+    # in ColorBar and possibly elsewhere
+
+    def __init__(self, ids):
+        """Give a list of unique IDs that will be remapped to their position in
+        this list.
+        """
+        assert ids.size > 0
+        fltids = np.array(ids, dtype=float)
+        bounds = np.concatenate([fltids - 0.5, [fltids[-1] + 0.5]])
+        super().__init__(bounds, len(bounds) - 1, clip=False)
+        self.vmin = 0
+        self.vmax = ids[-1]
+
+        # Remap matids to ID index: array up to the last valid matid.
+        ids_to_idx = np.zeros((ids[-1] + 1), dtype=np.int16)
+        ids_to_idx[ids] = np.arange(len(ids), dtype=np.int16)
```

Comment:
Does the int16 limit this to ~32k volumes? Not sure how crazy detector geometries are but I've seen CSG geometries in other domains with > 100k volumes.

Reply:
Oh right, actually since we're already remapping IDs into a zero-through-N list, I think this number is "IDs that have been plotted up to this point".

Reply:
Actually it looks like matplotlib has changed; it does an int->float mapping internally that means this branch is never called.

Reply:
And it seems like this specialization actually messes up the rendering sometimes... I'll delete it and if we have any performance issues later we can revisit.